### PR TITLE
[linux-6.6.y] Support for MIPS-based Loongson Devices

### PR DESCRIPTION
**Pull request by MingcongBai**
Backport most patches from the old Lemote tree to linux-6.6.y. This kernel is tested on 3A4000 and 3B4000 platforms.

### CODE REVIEW

File: drivers/cpufreq/Makefile [hunk 1]
```diff
@@ -102,9 +102,8 @@ obj-$(CONFIG_POWERNV_CPUFREQ)		+= powernv-cpufreq.o
 ##################################################################################
 # Other platform drivers
 obj-$(CONFIG_BMIPS_CPUFREQ)		+= bmips-cpufreq.o
-obj-$(CONFIG_IA64_ACPI_CPUFREQ)		+= ia64-acpi-cpufreq.o
```
Comment (opsiff) [inline drivers/cpufreq/Makefile:105]: 这几行多删了 会导致编译不出删掉的这几个驱动

**Review by opsiff — CHANGES_REQUESTED** [review]

File: drivers/cpufreq/Makefile (hunk 1 above)
Comment (MingcongBai) [reply drivers/cpufreq/Makefile:105]: 已修复

**Review by opsiff — APPROVED** [review]
/lgtm

File: drivers/cpufreq/Makefile [hunk 2]
```diff
@@ -105,6 +105,8 @@ obj-$(CONFIG_BMIPS_CPUFREQ)		+= bmips-cpufreq.o
 obj-$(CONFIG_IA64_ACPI_CPUFREQ)		+= ia64-acpi-cpufreq.o
 obj-$(CONFIG_LOONGSON2_CPUFREQ)		+= loongson2_cpufreq.o
 obj-$(CONFIG_LOONGSON3_ACPI_CPUFREQ)	+= loongson3-acpi-cpufreq.o
+obj-$(CONFIG_GS464_CPUFREQ)		+= gs464_cpufreq.o
+obj-$(CONFIG_LOONGSON1_CPUFREQ)		+= loongson1-cpufreq.o
```
Comment (opsiff) [inline drivers/cpufreq/Makefile:109]: 这里多了一个冗余的loongson1-cpufreq.o 这行 建议删掉

File: drivers/scsi/mvsas/mv_sas.c [hunk 3]
```diff
@@ -45,12 +45,17 @@ static int mvs_tag_alloc(struct mvs_info *mvi, u32 *tag_out)
 	unsigned int index, tag;
 	void *bitmap = mvi->rsvd_tags;
 
-	index = find_first_zero_bit(bitmap, MVS_RSVD_SLOTS);
-	tag = index;
-	if (tag >= MVS_RSVD_SLOTS)
+	index = find_next_zero_bit(bitmap, mvi->tags_num, mvi->next_tag);
```
Comment (opsiff) [inline drivers/scsi/mvsas/mv_sas.c]: 2024-12-23T08:51:37.4094692Z drivers/scsi/mvsas/mv_sas.c: In function ‘mvs_tag_alloc’:
2024-12-23T08:51:37.4096549Z drivers/scsi/mvsas/mv_sas.c:48:47: error: ‘struct mvs_info’ has no member named ‘tags_num’
2024-12-23T08:51:37.4097774Z    48 |         index = find_next_zero_bit(bitmap, mvi->tags_num, mvi->next_tag);
2024-12-23T08:51:37.4098431Z       |                                               ^~
2024-12-23T08:51:37.4099350Z drivers/scsi/mvsas/mv_sas.c:49:29: error: ‘struct mvs_info’ has no member named ‘tags_num’
2024-12-23T08:51:37.4100145Z    49 |         index = (index < mvi->tags_num)?
2024-12-23T08:51:37.4100618Z       |                             ^~
2024-12-23T08:51:37.4101464Z drivers/scsi/mvsas/mv_sas.c:50:62: error: ‘struct mvs_info’ has no member named ‘tags_num’
2024-12-23T08:51:37.4102397Z    50 |                         index:find_first_zero_bit(bitmap, mvi->tags_num);
2024-12-23T08:51:37.4102978Z       |                                                              ^~
2024-12-23T08:51:37.4103909Z drivers/scsi/mvsas/mv_sas.c:52:25: error: ‘struct mvs_info’ has no member named ‘tags_num’
2024-12-23T08:51:37.4104681Z    52 |         if (index >= mvi->tags_num)
2024-12-23T08:51:37.4105137Z       |                         ^~
2024-12-23T08:51:37.4105988Z drivers/scsi/mvsas/mv_sas.c:58:40: error: ‘struct mvs_info’ has no member named ‘tags_num’
2024-12-23T08:51:37.4106739Z    58 |         mvi->next_tag = (tag + 1) % mvi->tags_num;
2024-12-23T08:51:37.4107214Z       |                                        ^~
2024-12-23T08:51:37.4115883Z drivers/scsi/mvsas/mv_sas.c: In function ‘mvs_slave_configure’:
2024-12-23T08:51:37.4120314Z drivers/scsi/mvsas/mv_sas.c:275:31: error: unused variable ‘dev’ [-Werror=unused-variable]
2024-12-23T08:51:37.4122929Z   275 |         struct domain_device *dev = sdev_to_domain_dev(sdev);
2024-12-23T08:51:37.4125247Z       |                               ^~~
2024-12-23T08:51:37.4413491Z   CC [M]  net/smc/smc_ib.o
2024-12-23T08:51:37.4974509Z cc1: all warnings being treated as errors
2024-12-23T08:51:37.5097580Z make[5]: *** [scripts/Makefile.build:243：drivers/scsi/mvsas/mv_sas.o] 错误 1
2024-12-23T08:51:37.5104448Z make[4]: *** [scripts/Makefile.build:480：drivers/scsi/mvsas] 错误 2
2024-12-23T08:51:37.5105474Z make[3]: *** [scripts/Makefile.build:480：drivers/scsi] 错误 2
2024-12-23T08:51:37.5106571Z make[3]: *** 正在等待未完成的任务....

x86上编译不通过

**Review by opsiff — CHANGES_REQUESTED** [review]

File: drivers/scsi/mvsas/mv_sas.c (hunk 3 above)
Comment (opsiff) [reply drivers/scsi/mvsas/mv_sas.c]: 分析是6.2上面这个系列补丁删除掉了tags_num、tags变量
https://git.kernel.org/pub/scm/linux/kernel/git/stable/linux.git/commit/?h=v6.2&id=2acf97f199f9eba8321390325519e9b6bff60108

File: drivers/gpu/drm/radeon/radeon_pm.c [hunk 4]
```diff
@@ -1353,7 +1353,7 @@ static int radeon_pm_init_old(struct radeon_device *rdev)
 {
 	int ret;
 
-	rdev->pm.profile = PM_PROFILE_DEFAULT;
```
Comment (opsiff) [inline drivers/gpu/drm/radeon/radeon_pm.c:1356]: 这里建议用架构宏进行限制。

File: arch/mips/configs/deepin_loongson3_desktop_defconfig [hunk 5]
```diff
@@ -0,0 +1,406 @@
+# CONFIG_LOCALVERSION_AUTO is not set
```
Comment (opsiff) [inline arch/mips/configs/deepin_loongson3_desktop_defconfig:1]: 作为备注 后续需要同步其他架构的外设配置以支持广泛的外设

File: arch/mips/loongson64/env.c [hunk 6]
```diff
@@ -175,8 +182,14 @@ void __init prom_lefi_init_env(void)
 
 	loongson_sysconf.workarounds |= esys->workarounds;
 
+	mips_cpu_frequency = cpu_clock_freq;
 	pr_info("CpuClock = %u\n", cpu_clock_freq);
 
+	/* Append default cpu frequency with round-off */
+	sprintf(freq, " @ %uMHz", (cpu_clock_freq + 500000) / 1000000);
+	strncat(cpu_full_name, freq, sizeof(cpu_full_name));
```
Comment (Avenger-285714) [inline arch/mips/loongson64/env.c]: clang-19存在如下报错：
arch/mips/loongson64/env.c:190:31: error: the value of the size argument in 'strncat' is too large, might lead to a buffer overflow [-Werror,-Wstrncat-size]
  190 |         strncat(cpu_full_name, freq, sizeof(cpu_full_name));
      |                                      ^~~~~~~~~~~~~~~~~~~~~
arch/mips/loongson64/env.c:190:31: note: change the argument to be the free space in the destination buffer minus the terminating null byte
  190 |         strncat(cpu_full_name, freq, sizeof(cpu_full_name));
      |                                      ^~~~~~~~~~~~~~~~~~~~~
      |                                      sizeof(cpu_full_name) - strlen(cpu_full_name) - 1
1 error generated.

可改为：
strncat(cpu_full_name, freq, sizeof(cpu_full_name) - strlen(cpu_full_name) - 1);

**Comment by opsiff** [comment]
从逻辑上来看 应该修改为
strncat(cpu_full_name, freq, strlen(freq));



---Original---
From: ***@***.***&gt;
Date: Tue, Dec 24, 2024 14:46 PM
To: ***@***.***&gt;;
Cc: ***@***.******@***.***&gt;;
Subject: Re: [deepin-community/kernel] [linux-6.6.y] Support for MIPS-basedLoongson Devices (PR #528)




 
@Avenger-285714 commented on this pull request.
 
 
In arch/mips/loongson64/env.c:
 &gt;  	pr_info("CpuClock = %u\n", cpu_clock_freq);   +	/* Append default cpu frequency with round-off */ +	sprintf(freq, " @ %uMHz", (cpu_clock_freq + 500000) / 1000000); +	strncat(cpu_full_name, freq, sizeof(cpu_full_name));  
clang-19存在如下报错：
 arch/mips/loongson64/env.c:190:31: error: the value of the size argument in 'strncat' is too large, might lead to a buffer overflow [-Werror,-Wstrncat-size]
 190 |         strncat(cpu_full_name, freq, sizeof(cpu_full_name));
 |                                      ^~~~~~~~~~~~~~~~~~~~~
 arch/mips/loongson64/env.c:190:31: note: change the argument to be the free space in the destination buffer minus the terminating null byte
 190 |         strncat(cpu_full_name, freq, sizeof(cpu_full_name));
 |                                      ^~~~~~~~~~~~~~~~~~~~~
 |                                      sizeof(cpu_full_name) - strlen(cpu_full_name) - 1
 1 error generated.
 
可改为：
 strncat(cpu_full_name, freq, sizeof(cpu_full_name) - strlen(cpu_full_name) - 1);

**Pull request merged**